src/lib/IDBDAO.js

```diff
@@ -0,0 +1,350 @@
+        var os = tx.objectStore(self.name);
+        self.__TXN__[0] = os;
+        fn.call(self, os);
+      });
+    },
+
+    function put(value) {
+      var self = this;
+      return new Promise(function(resolve, reject) {
+        self.withStore("readwrite", function(store) {
+          var request = store.put(self.serialize(value), value.id);
+          request.transaction.addEventListener(
+            'complete',
+            function(e) {
+              self.pub('on','put', value);
+              resolve(value);
+            });
+          request.transaction.addEventListener(
+            'error',
+            function(e) {
+              reject(self.IDBInternalException.create({ id: value.id, error: e }));
+            });
+        });
+      });
+    },
+
+    function find(key) {
+      var self = this;
+
+      return new Promise(function(resolve, reject) {
+        self.withStore("readonly", function(store) {
+          var request = store.get(key);
+          request.transaction.addEventListener(
+            'complete',
+            function() {
+              if (!request.result) {
+                reject(self.ObjectNotFoundException.create({ id: key }));
+                return;
+              }
+              var result = self.deserialize(request.result);
+              resolve(result);
+            });
+          request.onerror = function(e) {
+            reject(self.IDBInternalException.create({ id: key, error: e }));
+          };
+        });
+      });
+    },
+
+    function remove(obj) {
+      var self = this;
+      var key = obj.id != undefined ? obj.id : obj;
+      return new Promise(function(resolve, reject) {
+        self.withStore("readwrite", function(store) {
+          var getRequest = store.get(key);
+          getRequest.onsuccess = function(e) {
+            if (!getRequest.result) {
+              // not found? as good as removed!
+              self.pub('on','remove', data);
+              resolve();
+              return;
+            }
+            var data = self.deserialize(getRequest.result);
+            var delRequest = store.delete(key);
+            delRequest.transaction.addEventListener('complete', function(e) {
+              self.pub('on','remove', data);
+              resolve();
+            });
+
+            delRequest.onerror = function(e) {
+              reject(self.IDBInternalException.create({ id: key, error: e }));
+            };
+          };
+          getRequest.onerror = function(e) {
+            reject(self.IDBInternalException.create({ id: key, error: e }));
+          };
+        });
+      });
+    },
+
+    function removeAll(sink, skip, limit, order, predicate) {
+      var query = predicate || this.True.create();
+
+      var self = this;
+
+      // If the caller doesn't care to see the objects as they get removed,
+      // then just nuke them in one go.
+      if ( ! predicate && ! ( sink && sink.remove ) ) {
```

Comment:
We need to check if anyone is listening for remove events.  If nobody is listening, and we have no predicate, then its safe to .clear().